File: src/Components/HmacSignatureProvider.cs
using HmacManager.Extensions;
using HmacManager.Headers;
using HmacManager.Mvc.Extensions.Internal;

namespace HmacManager.Components;

/// <summary>
/// Provides functionality to compute HMAC signatures and signing content for HTTP requests.
/// </summary>
public class HmacSignatureProvider : IHmacSignatureProvider
{
    /// <summary>
    /// Options used to configure the HMAC signature provider.
    /// </summary>
    protected readonly HmacSignatureProviderOptions Options;

    /// <summary>
    /// Initializes a new instance of the <see cref="HmacSignatureProvider"/> class with the specified options.
    /// </summary>
    /// <param name="options">The options for configuring the HMAC signature provider.</param>
    public HmacSignatureProvider(HmacSignatureProviderOptions options) => Options = options;

    /// <summary>
    /// Computes the HMAC signature for the provided signing content.
    /// </summary>
    /// <param name="signingContent">The content to sign.</param>
    /// <returns>A task that represents the asynchronous operation. The task result contains the computed HMAC signature as a string.</returns>
    /// <exception cref="ArgumentException">Thrown if <paramref name="signingContent"/> is null or whitespace.</exception>
    public Task<string> ComputeSignatureAsync(string signingContent)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(signingContent);
        return Options.SignatureHashGenerator.HashAsync(signingContent);
    }

    /// <summary>
    /// Computes the signing content for an HTTP request, including headers and content hash if applicable.
    /// </summary>
    /// <param name="request">The HTTP request message.</param>
    /// <param name="dateRequested">The date and time the request was made.</param>
    /// <param name="nonce">A unique nonce to include in the signing content.</param>
    /// <param name="headerValues">Optional collection of header values to include in the signing content.</param>
    /// <returns>A task that represents the asynchronous operation. The task result contains the computed signing content as a string.</returns>
    public async Task<string> ComputeSigningContentAsync(
        HttpRequestMessage request, 
        DateTimeOffset dateRequested, 
        Guid nonce,
        IReadOnlyCollection<HeaderValue>? headerValues = null
    )
    {
        var builder = Options.SigningContentBuilder.CreateBuilder()
            .WithRequest(request)
            .WithPublicKey(Options.Keys.PublicKey)
            .WithDateRequested(dateRequested)
            .WithNonce(nonce)
            .WithHeaderValues(headerValues ?? []);

        if (request.HasContent())
        {
            var contentString = await request.Content.ReadAsStringAsync();
            var contentHash = await Options.ContentHashGenerator.HashAsync(contentString);
            builder.WithContentHash(contentHash);
        }

        return builder.Build();
    }
}

File: client/lib/src/components/hmac-signature-provider.ts
import SignatureBuilder from "../builders/signature-builder"
import SigningContentBuilder from "../builders/signing-content-builder"
import HashAlgorithm from "../hash-algorithm";
import HmacSignature from "./hmac-signature";
import { computeContentHash } from "../utilities/hmac-utilities";

/**
 * Provides functionality to create an HMAC signature for request authentication.
 */
export default class HmacSignatureProvider {
    /** 
     * The public key used in the HMAC signing process.
     */
    private readonly publicKey: string;

    /** 
     * The private key used for signing the content in the HMAC process.
     */
    private readonly privateKey: string;

    /** 
     * A list of headers that are included in the HMAC signature.
     * Each string in the array represents a header that has been included.
     */
    private readonly signedHeaders: string[] = [];

    /** 
     * The hashing algorithm used to compute the content hash. 
     * Default is SHA-256.
     */
    private readonly contentHashAlgorithm: HashAlgorithm = HashAlgorithm.SHA256;

    /** 
     * The hashing algorithm used to compute the signature hash. 
     * Default is SHA-256.
     */
    private readonly signatureHashAlgorithm: HashAlgorithm = HashAlgorithm.SHA256;

    private readonly signingContentBuilder: SigningContentBuilder;

    /**
     * Initializes a new instance of HmacSignatureProvider.
     * @param publicKey - The public key used for signature generation.
     * @param privateKey - The private key used for signature generation.
     * @param signedHeaders - An array of headers to be signed. Defaults to an empty array.
     * @param contentHashAlgorithm - The algorithm used for content hashing. Defaults to "sha-256".
     * @param signatureHashAlgorithm - The algorithm used for signature hashing. Defaults to "sha-256".
     */
    constructor(
        publicKey: string,
        privateKey: string,
        signedHeaders: string[] = [],
        contentHashAlgorithm: HashAlgorithm = HashAlgorithm.SHA256,
        signatureHashAlgorithm: HashAlgorithm = HashAlgorithm.SHA256,
        signingContentBuilder: SigningContentBuilder = new SigningContentBuilder()
    ) {
        this.publicKey = publicKey;
        this.privateKey = privateKey;
        this.signedHeaders = signedHeaders;
        this.contentHashAlgorithm = contentHashAlgorithm;
        this.signatureHashAlgorithm = signatureHashAlgorithm;
        this.signingContentBuilder = signingContentBuilder;
    }

    /**
     * Computes the HMAC signature for a given request, using the specified date and nonce.
     * @param request - The request to be signed.
     * @param dateRequested - The timestamp of the request.
     * @param nonce - A unique identifier for the request.
     * @returns An object containing the signing content and generated signature.
     */
    compute = async (request: Request, dateRequested: Date, nonce: string): Promise<HmacSignature> => {
        // Clone the request since we need to read the body if it exists
        const { body } = request.clone();
        const contentHash = await computeContentHash(body, this.contentHashAlgorithm);
        const signingContentBuilder = this.signingContentBuilder.createBuilder()
            .withRequest(request)
            .withPublicKey(this.publicKey)
            .withDateRequested(dateRequested)
            .withContentHash(contentHash)
            .withSignedHeaders(this.signedHeaders)
            .withNonce(nonce);

        const signingContent = await signingContentBuilder.build();
        const signatureBuilder = new SignatureBuilder(this.privateKey,
            signingContent,
            this.signatureHashAlgorithm
        );

        return {
            signingContent,
            signature: await signatureBuilder.build()
        };
    }
}